hyperopt/spark.py

```diff
@@ -0,0 +1,415 @@
+from __future__ import print_function
+import numbers
+import threading
+import time
+import timeit
+
+import cloudpickle
+
+from hyperopt import base, fmin, Trials
+from hyperopt.utils import coarse_utcnow, _get_logger, _get_random_id
+from pyspark.sql import SparkSession
+
+
+logger = _get_logger('hyperopt-spark')
+
+
+class SparkTrials(Trials):
+    """
+    Implementation of hyperopt.Trials supporting
+    distributed execution using Apache Spark clusters.
+    This requires fmin to be run on a Spark cluster.
+
+    Plugging SparkTrials into hyperopt.fmin() allows hyperopt
+    to send model training and evaluation tasks to Spark workers,
+    parallelizing hyperparameter search.
+    Each trial (set of hyperparameter values) is handled within
+    a single Spark task; i.e., each model will be fit and evaluated
+    on a single worker machine.  Trials are run asynchronously.
+
+    See hyperopt.Trials docs for general information about Trials.
+
+    The fields we store in our trial docs match the base Trials class.  The fields include:
+     - 'tid': trial ID
+     - 'state': JOB_STATE_DONE, JOB_STATE_ERROR, etc.
+     - 'result': evaluation result for completed trial run
+     - 'refresh_time': timestamp for last status update
+     - 'misc': includes:
+        - 'error': (error type, error message)
+     - 'book_time': timestamp for trial run start
+    """
+
+    asynchronous = True
+
+    # Hard cap on the number of concurrent hyperopt tasks (Spark jobs) to run. Set at 128.
+    MAX_CONCURRENT_JOBS_ALLOWED = 128
```

Comment:
just curious, why 128? Might be a good idea to add util functionality to change this, if needed.

Reply:
It's pretty arbitrary.  We did a bit of stress testing and found this seemed fine; we got input from other teams who worried about pushing it closer to 1000; I believe Spark has a hard cap around 2000; and we want to do more stress testing in the future.

I'm up for making it a changeable config.  Do you mind if we make that a follow-up?  I started a list of such items at the end of the design doc: https://docs.google.com/document/d/1R-qMVOYxaLw3HdqMiTADQDZ5z7uInMR-x42HFCXu2-s/edit#heading=h.kqrwh3pgxu44